File: compute/pyo3/python/mog/sub_apis/protection.py
"""Sheet protection APIs that are not production-backed in Python yet."""
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, Optional, Union

from mog._unsupported import UnsupportedSubApiMixin, unsupported_python_path

if TYPE_CHECKING:
    from mog._bridge import Bridge
    from mog.types import MutationResult


class ProtectionAPI(UnsupportedSubApiMixin):
    """Sheet protection operations on a worksheet."""

    __slots__ = ("_bridge", "_sheet_id_json")
    _UNSUPPORTED_ACCESSOR_API_PATH = "ws.protection"
    _UNSUPPORTED_ACCESSOR_PYTHON_PATH = "ws.protection"

    def __init__(self, bridge: Bridge, sheet_id_json: str) -> None:
        self._bridge = bridge
        self._sheet_id_json = sheet_id_json

    def mark_unlocked(self, row: int, col: int) -> None:
        unsupported_python_path("ws.protection.mark_unlocked")

    def mark_locked(self, row: int, col: int) -> None:
        unsupported_python_path("ws.protection.mark_locked")

    def protect(
        self,
        password: Optional[str] = None,
        options: Optional[Dict[str, Any]] = None,
    ) -> "MutationResult":
        unsupported_python_path("ws.protection.protect")

    def unprotect(self, password: Optional[str] = None) -> Union["MutationResult", bool]:
        unsupported_python_path("ws.protection.unprotect")

    def is_protected(self) -> bool:
        unsupported_python_path("ws.protection.is_protected")

    def can_edit_cell(self, row: int, col: int) -> bool:
        unsupported_python_path("ws.protection.can_edit_cell")

    def can_sort(self) -> bool:
        unsupported_python_path("ws.protection.can_sort")

    def can_do_structure_op(self, operation: str) -> bool:
        unsupported_python_path("ws.protection.can_do_structure_op")

    def get_config(self) -> Dict[str, Any]:
        unsupported_python_path("ws.protection.get_config")

    def get_selection_mode(self) -> str:
        unsupported_python_path("ws.protection.get_selection_mode")

    def set_selection_mode(self, mode: str) -> None:
        unsupported_python_path("ws.protection.set_selection_mode")

    @property
    def allow_edit_ranges(self) -> Any:
        return self.__getattr__("allow_edit_ranges")

File: types/api/src/api/worksheet/protection.ts
/**
 * WorksheetProtection — Sub-API Interface for Sheet Protection
 *
 * Methods for protecting/unprotecting sheets, querying protection state,
 * and checking whether specific operations are allowed under protection.
 */
import type { ProtectionConfig, ProtectionOptions } from '../types';

/** Valid operations for structural protection checks. */
export type ProtectionOperation =
  | 'insertRows'
  | 'insertColumns'
  | 'deleteRows'
  | 'deleteColumns'
  | 'formatCells'
  | 'formatRows'
  | 'formatColumns'
  | 'sort'
  | 'filter'
  | 'editObject';

/** A range that can be edited even when the sheet is protected. */
export interface AllowEditRange {
  /** Title/name of the editable range */
  title: string;
  /** Range address in A1 notation (e.g. "A1:C10") */
  address: string;
}

/** Collection of ranges that can be edited while the sheet is protected. */
export interface WorksheetAllowEditRanges {
  /**
   * Add a range that can be edited when the sheet is protected.
   *
   * @param title - Name for the editable range
   * @param address - Range address in A1 notation
   */
  add(title: string, address: string): Promise<void>;

  /**
   * Remove an allow-edit range by title.
   *
   * @param title - Title of the range to remove
   */
  remove(title: string): Promise<void>;

  /**
   * List all allow-edit ranges.
   *
   * @returns Array of allow-edit range definitions
   */
  list(): Promise<AllowEditRange[]>;
}

/** Sub-API for worksheet protection operations. */
export interface WorksheetProtection {
  /**
   * Check whether the sheet is currently protected.
   *
   * @returns True if the sheet is protected
   */
  isProtected(): Promise<boolean>;

  /**
   * Protect the sheet with an optional password.
   *
   * @param password - Optional password to require for unprotection
   */
  protect(password?: string): Promise<void>;

  /**
   * Protect the sheet with an optional password and granular permission options.
   *
   * @param password - Optional password to require for unprotection
   * @param options - Granular permissions (e.g., allowSort, allowFormatCells)
   */
  protectWithOptions(password?: string, options?: ProtectionOptions): Promise<void>;

  /**
   * Unprotect the sheet. Returns true if unprotection succeeded.
   *
   * @param password - Password if the sheet was protected with one
   * @returns True if the sheet was successfully unprotected
   */
  unprotect(password?: string): Promise<boolean>;

  /**
   * Check whether a specific cell can be edited under current protection.
   * Returns true if the sheet is unprotected or the cell is unlocked.
   *
   * @param row - Row index (0-based)
   * @param col - Column index (0-based)
   * @returns True if the cell is editable
   */
  canEditCell(row: number, col: number): Promise<boolean>;

  /**
   * Synchronous protection fast path for edit-entry commands.
   *
   * Returns true only when the mirror proves the sheet is currently
   * unprotected. Returns 'unknown' for protected sheets, where callers must use
   * canEditCell(row, col) so cell lock state remains kernel-owned.
   */
  canEditCellFast(row: number, col: number): true | 'unknown';

  /**
   * Check whether a structural operation is allowed under current protection.
   *
   * @param operation - Operation name (e.g., 'sort', 'insertRows', 'deleteColumns')
   * @returns True if the operation is allowed
   */
  canDoStructureOp(operation: ProtectionOperation): Promise<boolean>;

  /**
   * Check whether sorting is allowed under current protection.
   * Convenience shorthand for `canDoStructureOp('sort')`.
   *
   * @returns True if sorting is allowed
   */
  canSort(): Promise<boolean>;

  /**
   * Get the full protection configuration for the sheet.
   *
   * @returns Protection configuration including status and permission flags
   */
  getConfig(): Promise<ProtectionConfig>;

  /**
   * Get the current selection mode when the sheet is protected.
   *
   * @returns 'normal' (all cells selectable), 'unlockedOnly' (only unlocked cells), or 'none' (no selection)
   */
  getSelectionMode(): Promise<'normal' | 'unlockedOnly' | 'none'>;

  /**
   * Set the selection mode for when the sheet is protected.
   * Only takes effect when the sheet is actually protected.
   *
   * @param mode - Selection mode
   */
  setSelectionMode(mode: 'normal' | 'unlockedOnly' | 'none'): Promise<void>;

  /**
   * Temporarily suspend sheet protection.
   * If the sheet is password-protected, the correct password must be provided.
   *
   * @param password - Password if the sheet is password-protected
   */
  pauseProtection(password?: string): Promise<void>;

  /**
   * Re-enable previously paused protection with the original configuration.
   */
  resumeProtection(): Promise<void>;

  /**
   * Whether protection can be paused (true if sheet is protected and not already paused).
   */
  readonly canPauseProtection: boolean;

  /**
   * Whether protection is currently paused.
   */
  readonly isPaused: boolean;

  /**
   * Check if the given password matches the current protection password
   * without modifying protection state.
   *
   * @param password - Password to check
   * @returns True if the password matches
   */
  checkPassword(password?: string): Promise<boolean>;

  /**
   * Change the protection password without needing to unprotect/re-protect.
   *
   * @param password - New password, or undefined/empty to remove the password
   */
  setPassword(password?: string): Promise<void>;

  /**
   * Update protection options without needing to unprotect/re-protect.
   *
   * @param options - Partial protection options to merge with existing options
   */
  updateOptions(options: Partial<ProtectionOptions>): Promise<void>;

  /**
   * Collection of ranges that can be edited even when the sheet is protected.
   */
  readonly allowEditRanges: WorksheetAllowEditRanges;
}
